src/axom/quest/Delaunay.hpp

```diff
@@ -0,0 +1,885 @@
+    int element_index) const
+  {
+    const auto verts = m_mesh.boundaryVertices(element_index);
+    const PointType& p0 = m_mesh.getVertexPosition(verts[0]);
+    const PointType& p1 = m_mesh.getVertexPosition(verts[1]);
+    const PointType& p2 = m_mesh.getVertexPosition(verts[2]);
+    const PointType& p3 = m_mesh.getVertexPosition(verts[3]);
+
+    return ElementType(p0, p1, p2, p3);
+  }
+
+  /**
+   * \brief Prints out mesh details, for debugging purpose.
+   */
+  void printMesh() { m_mesh.print_all(); }
+
+  /**
+   * \brief Write the m_mesh to a legacy VTK file
+   *
+   * \param filename The name of the file to write to,
+   * \note The suffix ".vtk" will be appended to the provided filename
+   * \details This function uses mint to write the m_mesh to VTK format.
+   */
+  void writeToVTKFile(const std::string& filename)
+  {
+    const auto CELL_TYPE = DIM == 2 ? mint::TRIANGLE : mint::TET;
+    mint::UnstructuredMesh<mint::SINGLE_SHAPE> mint_mesh(DIM, CELL_TYPE);
+
+    this->compactMesh();
+
+    for(auto v : m_mesh.vertices().positions())
+    {
+      mint_mesh.appendNodes(m_mesh.getVertexPosition(v).data(), 1);
+    }
+
+    for(auto e : m_mesh.elements().positions())
+    {
+      mint_mesh.appendCell(&(m_mesh.boundaryVertices(e)[0]), CELL_TYPE);
+    }
+
+    mint::write_vtk(&mint_mesh, filename);
+  }
+
+  /**
+   * \brief Removes the vertices that defines the boundary of the mesh,
+   * and the elements attached to them.
+   *
+   * \details After this function is called, no more points can be added to the m_mesh.
+   */
+  void removeBoundary()
+  {
+    if(m_has_boundary)
+    {
+      //remove the boundary box, which will be the first 4 points for triangles, first 8 for tetrahedron
+      const int num_boundary_pts = 1 << DIM;
+
+      //Collect a list of elements to remove first, because
+      //the list may be incomplete if generated during the removal.
+      IndexArray elements_to_remove;
+      for(int v = 0; v < num_boundary_pts; ++v)
+      {
+        IndexArray elems = m_mesh.vertexStar(v);
+        elements_to_remove.insert(elements_to_remove.end(),
+                                  elems.begin(),
+                                  elems.end());
+      }
+      for(auto e : elements_to_remove)
+      {
+        if(m_mesh.isValidElement(e))
+        {
+          m_mesh.removeElement(e);
+        }
+      }
+
+      for(int v = 0; v < num_boundary_pts; ++v)
+      {
+        m_mesh.removeVertex(v);
+      }
+
+      this->compactMesh();
+      m_has_boundary = false;
+    }
+  }
+
+  /// \brief Get the IA mesh data pointer
+  const IAMeshType* getMeshData() const { return &m_mesh; }
+
+  /**
+   * \brief Checks that the underlying mesh is a valid Delaunay triangulation of the point set
+   *
+   * A Delaunay triangulation is valid when none of the vertices are inside the circumspheres 
+   * of any of the elements of the mesh
+   */
+  bool isValid(bool verboseOutput = false) const
```

Comment:
The validity check is using an entirely different algorithm to check for validity.
* Insert points uses the mesh connectivity to find and remove invalid elements using the (implicit) in_sphere() algorithm.
* isValid() builds up a spatial index over the (explicit) circumspheres of each element and then uses `Sphere::getOrientation()`